qiskit/pulse/builder.py

```diff
@@ -769,39 +784,46 @@ def call_subroutine(
             self._compile_lazy_circuit()
             subroutine = self._compile_circuit(subroutine)
 
-        empty_subroutine = True
-        if isinstance(subroutine, Schedule):
-            if len(subroutine.instructions) > 0:
-                empty_subroutine = False
-        elif isinstance(subroutine, ScheduleBlock):
-            if len(subroutine.blocks) > 0:
-                empty_subroutine = False
-        else:
+        if not isinstance(subroutine, (Schedule, ScheduleBlock)):
             raise exceptions.PulseError(
                 f"Subroutine type {subroutine.__class__.__name__} is "
                 "not valid data format. Call QuantumCircuit, "
                 "Schedule, or ScheduleBlock."
             )
 
-        if not empty_subroutine:
-            param_value_map = {}
-            for param_name, assigned_value in kw_params.items():
-                param_objs = subroutine.get_parameters(param_name)
-                if len(param_objs) > 0:
-                    for param_obj in param_objs:
-                        param_value_map[param_obj] = assigned_value
-                else:
-                    raise exceptions.PulseError(
-                        f"Parameter {param_name} is not defined in the target subroutine. "
-                        f'{", ".join(map(str, subroutine.parameters))} can be specified.'
-                    )
-
-            if value_dict:
-                param_value_map.update(value_dict)
+        if len(subroutine) == 0:
+            return
 
-            call_def = instructions.Call(subroutine, param_value_map, name)
-
-            self.append_instruction(call_def)
+        # Create local parameter assignment
+        local_assignment = dict()
+        for param_name, value in kw_params.items():
+            params = subroutine.get_parameters(param_name)
+            if not params:
+                raise exceptions.PulseError(
+                    f"Parameter {param_name} is not defined in the target subroutine. "
+                    f'{", ".join(map(str, subroutine.parameters))} can be specified.'
+                )
+            for param in params:
+                local_assignment[param] = value
+        if value_dict:
+            local_assignment.update(value_dict)
+
+        if isinstance(subroutine, ScheduleBlock):
+            # If subroutine is schedule block, use reference mechanism.
+            if local_assignment:
+                subroutine = subroutine.assign_parameters(local_assignment, inplace=False)
+            if name is None:
+                # Add unique string, not to accidentally override existing reference entry.
+                keys = (subroutine.name, uuid.uuid4().hex)
+            else:
+                keys = (name,)
```

Comment:
I think always using shadowing is probably best. You need to use shadowing when the name is not given because a schedule could call multiple schedules with the same name (e.g. a two qubit schedule calling x on each qubit). While using the name as the key feels intuitive, it's not needed to preserve the way `call()` worked before.

One issue with the current behavior is that if you use the same name for multiple calls with different schedules (which does not seem like a good thing to do) the last one overwrites the earlier ones whereas before the individual schedules would be kept.

Besides backwards compatibility, the nice thing about `call()` is that adds a reference and assigns it in a single step. Maybe the new `reference()` method could optionally take a `ScheduleBlock` and assign it if given? Since it is a new method, it does not have to worry about backwards compatibility and can just always use the required `name` parameter as the key.

Reply:
I think we used opposite meanings of "shadowing" here - I meant "overwrites" (so the previous definitions of "x" are inaccessible now), and I think you meant "adds to"?  It doesn't matter too much if we did - I don't have any position on what's best for the API here, because I've never used the code.  As long as you and Naoki come to an agreement on what's best, I can sign off on the PR.  If that's the extra uuid bit of the key in all cases, that's absolutely fine by me.

Reply:
Yes, you are right. I got it backwards from you. I was thinking of the function obscuring the subroutine's name as shadowing the name.

Reply:
This is good point. I'm not 100 percent confident on the current implementation. This at least guarantees the conventional behavior. The reference key is important only when a user assigns a subroutine to reference. Since, `.call` performs this operation in a single function, basically the reference key can be whatever. There is no reason to update explicitly provided `name`, but I added uuid here to avoid _local parameter collision_ since we can call `.call` with parameter assignment. Remember that reference manager does NOT allow assigning multiple objects to a single reference. Calling the same schedule with different parameter assignments will generate different subroutines with the same name on the fly, and this operation usually fails without shadowing. This should be supported for backward compatibility. I wanted to generate unique name considering parameters, but I needed to give up because of mutability of `.assign_parameters` (I needed unnecessarily heavy logic to realize). Another option would be 
```suggestion
            keys = name or uuid.uuid4().hex
```
probably this makes more sense since `subroutine.name` is no longer the effective identifier due to uuid.

Reply:
> The reference key is important only when a user assigns a subroutine to reference.

The key also factors into the new parameter scoping features. If you call a schedule without passing a name, you will get difficult to use scoped parameter names.

One thing I wanted to check -- do you want to support calling different schedules with the same name passed explicitly? It sounds like the reference manager will raise an exception for the second schedule now whereas before the PR it was allowed to use the same name since the name was just attached to the instruction and not inserted into a shared namespace.

> keys = name or uuid.uuid4().hex

The problem with this is that the first key gets used for `Instruction.name` so it is probably best to keep that a readable string.

Reply:
Yes, this changes the behavior. I updated API doc and release note accordingly in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/ffd2130f01b0ad9e0bcdc0a88d716b8104f01b2f

> The problem with this is that the first key gets used for Instruction.name so it is probably best to keep that a readable string.

These comments all make sense. I leave the current implementation.